docs/proposals/node-attribute-labels.md

```diff
@@ -0,0 +1,119 @@
+such as SSD and GPU are consumable and should be represented as generic resources.
+See [generic resources](../../docs/design/resources.md).
+
+## Proposal
+
+### Label Namespace
+
+Node attribute labels will live under the namespace `<some prefix>.node.kubernetes.io` and be required
+to match the glob `*.node.kubernetes.io/*`.  The name space `node.kubernetes.io` is reserved for use by
+kubernetes.
+
+### Plugins
+
+We will allow cloud providers and kubernetes administrators to specify custom node attribute
+labels through plugins.  Plugins will be checked at node start up time and published
+to the api server as part of the Node.ObjectMeta.Labels structure.  The plugin directory
+can be specified via the kubelet server command line argument `node-label-plugin-dir`.
+This flag specifies a directory containing one or more executables or json files.  This directory
+should NOT be mountable by containers via HostPath.  At kubelet initialization time these files
+will be used to apply additional labels to the node.  After start up, plugins will be polled periodically
+(every 10 minutes) in a background thread to update the labels sent as part of Node.ObjectMeta.Labels
+during health checks.  If there are errors while updating the labels, then no updates will be applied
+and the last good set will be used.  If there is an error during the first time plugins are checked,
+then no additional labels will be applied but the non-plugin specified node labels will still be used.
+
+## Plugin File Format
+
+### Json
+
+Json file plugins will be identified by having the extension '.json' and expected to be a map[string]string.
+e.g. a file named `label.json` with contents`{"a.node.kubernetes.io/l1":"v1", "a.node.kubernetes.io/l2":"v2"}`
+
+## Executable
+
+Executable file plugins will be identified by having a white listed extension.  The initial whitelist will
+be ['.sh', '.py'].  Executable file plugins are expected to be executable by the kubelet.  If the
+kubelet cannot execute a plugin identified by its extension, it is considered an error and no labels
+will be updated.  Executable plugins are expected to emit to stdout json that can be parsed as a json plugin.
+e.g. a file name `label.sh` with contents `echo '{"a.node.kubernetes.io/l1":"v1", "a.node.kubernetes.io/l2":"v2"}'`
+
+### Export of Labels to Api Server
+
+Labels will be exported to the Api server as part of the api.Node.ObjectMeta.Labels map at registration
+and during health checks.
+
+## Other Considerations
+
+- Node labels are not versioned and cannot easily be changed.  We should be careful about the labels we apply.
+- Label updates are done via periodic polling.  There are more efficient ways of getting updates at the cost of
+additional complexity (e.g. how do we get them from executable plugins?).
+
+## Open questions
+
+- __Do we really need to support executables as a plugin format?  Is supporting json files only sufficient?__
```

Comment:
I feel like files aligns pretty closely with static pod manifests that we already have - doesn't seem bad to support both.